api/src/main/java/javax/config/ComposedConfig.java

```diff
@@ -0,0 +1,254 @@
+ * <p>
+ * The hierarchical resolution works both for implicitly resolved fields and explicitly annotated members.
+ *
+ * <pre>
+ * &#064;ComposedConfig
+ * public class ServerConfig {
+ *
+ *     private String host;
+ *     private SocketConfig socket;
+ *
+ *     // getters & setters
+ *
+ *     public static class SocketConfig {
+ *         private String name;
+ *         private String protocol;
+ *         private int port;
+ *
+ *         // getters & setters
+ *     }
+ *
+ * }
+ * </pre>
+ * <p>
+ * If a {@code ServerConfig} configuration type is retrieved, the property keys are resolved as follows:
+ *
+ * <pre>
+ * public class SomeBean {
+ *
+ *     &#064;Inject
+ *     &#064;ConfigProperty(name = "server")
+ *     private ServerConfig serverConfig;
+ *
+ * }
+ * </pre>
+ * <p>
+ * This leads to:
+ *
+ * <pre>
+ * server.host
+ * server.socket.name
+ * server.socket.protocol
+ * server.socket.port
+ * </pre>
+ * <p>
+ * The property keys are resolved by the field names, or the names defined in {@link ConfigProperty}, respectively, and combined via dot ({@code .}).
+ * <p>
+ * The example above is congruent with annotating {@code SocketConfig} with {@link ComposedConfig}, as well.
+ *
+ * <h3>Collection resolution</h3>
+ * <p>
+ * Composed configuration types also resolve collections and array types.
+ *
+ * <pre>
+ * &#064;ComposedConfig
+ * public class MultiSocketServerConfig {
+ *
+ *     private String[] hosts;
+ *     private List<SocketConfig> sockets;
+ *
+ *     // getters & setters
+ *
+ *     public static class SocketConfig {
+ *         private String name;
+ *         private String protocol;
+ *         private int port;
+ *
+ *         // getters & setters
+ *     }
+ *
+ * }
+ * </pre>
+ * <p>
+ * If the {@code MultiSocketServerConfig} type is resolved by key {@code alternative-server}, it results in the following:
+ *
+ * <pre>
+ * server.hosts.0
+ * server.hosts.1
+ *
+ * server.sockets.0.name
+ * server.sockets.0.protocol
+ * server.sockets.1.name
+ * </pre>
+ * <p>
+ * Element types of collections and arrays are resolved by an implicit zero-based index, which is part of the resulting, combined property key.
+ * <p>
+ * This collection resolution works for array types, and types that are assignable to {@link java.util.Collection}.
+ * For unordered collection types, e.g. {@link java.util.Set}, the order in which the configured elements will be retrieved is non-deterministic, despite the (zero-based) indexed key names.
+ * <p>
+ * Similar to singular sub-types, the element types within the collection or array are resolved by potentially existent converters, and resolved recursively if no built-in, custom, or implicit converters are defined.
+ *
+ * @author <a href="mailto:mail@sebastian-daschner.com">Sebastian Daschner</a>
+ */
+@Qualifier
+@Retention(RUNTIME)
+@Target({METHOD, FIELD, PARAMETER, TYPE})
+public @interface ComposedConfig {
+
+    /**
```

Comment:
probably have an attribute 'prefix' or 'name'?
In that case you could write 
```
@ComposedConfig(prefix="com.acme.my.server1"
public interface MultiSocketServerConfig {
  String hostname();
  Integer port();
```
which will result in taking the method names as config attribute name. In this case this would result in `com.acme.my.server1.hostname` and `com.acme.my.server1.port`.

Reply:
Oki, not needed as @sdaschner explained in the EG meeting. The prefix is defined at the point where you use it e.g. with `@Inject @ConfigProperty(name="com.acme.my.server1")`